src/tactic/hint.lean

```diff
@@ -66,23 +66,33 @@ add_hint_tactic "unfold_aux"
 
 end hint
 
-/-- report a list of tactics that can make progress against the current goal -/
-meta def hint : tactic (list string) :=
-do names ← attribute.get_instances `hint_tactic,
-   try_all_sorted (names.reverse.map name_to_tactic)
+/--
+Report a list of tactics that can make progress against the current goal,
+and for each such tactic, the number of remaining goals afterwards.
+-/
+meta def hint : tactic (list (string × ℕ)) :=
+do
+  names ← attribute.get_instances `hint_tactic,
+  focus1 $ try_all_sorted (names.reverse.map name_to_tactic)
 
 namespace interactive
 
 /--
-report a list of tactics that can make progress against the current goal
+Report a list of tactics that can make progress against the current goal.
 -/
 meta def hint : tactic unit :=
-do hints ← tactic.hint,
-   if hints.length = 0 then
-     fail "no hints available"
-   else
-     do trace "the following tactics make progress:\n----",
-        hints.mmap' (λ s, tactic.trace format!"Try this: {s}")
+do
+  hints ← tactic.hint,
+  if hints.length = 0 then
+    fail "no hints available"
+  else do
+    t ← hints.nth 0,
+    if t.2 = 0 then do
+      trace "the following tactics solve the goal:\n----",
```

Comment:
What happens if some tactics solve the goal, but others just make progress?

Reply:
Then it only lists goals that actually close the goal. Which seems fine. I don't think I would be interested in the "long tail" of the list that only makes progress if there is a tactic that actually closes the goal for me.

Reply:
Oh, I should have read the line below this one. :smile: